### PR TITLE
Try to deal with calling .wait()

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90877
* #90813
* #90804

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -1117,23 +1117,24 @@ def codegen(self):
                     or node.is_template()
                 ):
                     self.flush()
+                    print(f"Switching from {self.current_device} to {device}")
                     self.current_device = device
+            with torch.cuda.device(self.current_device):
```
Comment (ngimel) [inline torch/_inductor/scheduler.py:1122]: I don't think you should be setting current_device here, instead you should be emitting this code in the codegen_wrapper (and this unfortunately would require threading the initial indent through all IndentedBuffers)

Comment (wconstab) [reply torch/_inductor/scheduler.py:1122]: ok- will try to do that. thanks!

Comment (wconstab) [reply torch/_inductor/scheduler.py:1122]: removing this for now and will intercept https://github.com/pytorch/pytorch/pull/90934

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -3404,54 +3404,55 @@ def get_template_tiling(self):
...
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(layout, inputs, constant_args)
 
-#     def get_mutation_names(self):
-#         assert isinstance(self.layout, MutationLayout)
-#         return (self.layout.target.get_name(),)
+    def codegen(self, wrapper):
+        (work,) = self.constant_args
```
Comment (jansel) [inline torch/_inductor/ir.py:3417]: Should this be a normal input (not a constant input)?  I worry there won't be a dependency edge in the scheduler that forces this to run after the allreduce.

You might also want to create fake dependencies to ensure ordering.

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -3404,54 +3404,55 @@ def get_template_tiling(self):
...
     ):
         x = cls.realize_input(x)
-        return AllReduce(
+        all_reduce = AllReduce(
             layout=MutationLayout(x),
             inputs=[x],
         )
+        realized_allreduce = cls.realize_input(all_reduce)
+        return Wait(
+            layout=FlexibleLayout(
```
Comment (jansel) [inline torch/_inductor/ir.py:3448]: FixedLayout?

**Pull request closed without merging**